storage/bootstrap/bootstrapper/peers/source.go

```diff
@@ -188,79 +168,152 @@ func (s *peersSource) Read(
-		}
-
-		if err := blockRetriever.CacheShardIndices(shards); err != nil {
+		err := s.cacheShardIndices(shardsTimeRanges, blockRetriever)
+		if err != nil {
 			return nil, err
 		}
 	}
 
 	return result, nil
 }
 
+// startIncrementalQueueWorker is meant to be run in its own goroutine, and it creates a worker that
+// loops through the incrementalQueue and performs an incrementalFlush for each entry, ensuring that
+// no more than one incremental flush is ever happening at once. Once the incrementalQueue channel
+// is closed, and the worker has completed flushing all the remaining entries, it will close the
+// provided doneCh so that callers can block until everything has been successfully flushed.
+func (s *peersSource) startIncrementalQueueWorker(
+	doneCh chan struct{},
+	incrementalQueue chan incrementalFlush,
+	persistFlush persist.Flush,
+	bootstrapResult result.BootstrapResult,
+	lock *sync.Mutex,
+) {
+	// If performing an incremental bootstrap then flush one
+	// at a time as shard results are gathered
+	for flush := range incrementalQueue {
+		err := s.incrementalFlush(persistFlush, flush.nsMetadata, flush.shard,
+			flush.shardRetrieverMgr, flush.shardResult, flush.timeRange)
+		if err == nil {
+			// Safe to add to the shared bootstrap result now
+			lock.Lock()
+			bootstrapResult.Add(flush.shard, flush.shardResult, xtime.Ranges{})
+			lock.Unlock()
+			continue
+		}
+
+		// Remove results and make unfulfilled if an error occurred
+		s.log.WithFields(
+			xlog.NewField("error", err.Error()),
+		).Errorf("peers bootstrapper incremental flush encountered error")
+
+		// Make unfulfilled
+		lock.Lock()
+		bootstrapResult.Add(flush.shard, nil, xtime.Ranges{}.AddRange(flush.timeRange))
+		lock.Unlock()
+	}
+	close(doneCh)
+}
+
+// fetchBootstrapBlocksFromPeers loops through all the provided ranges for a given shard and
+// fetches all the bootstrap blocks from the appropriate peers.
+// 		Non-incremental case: Immediately add the results to the bootstrap result
+// 		Incremental case: Don't add the results yet, but push an incrementalFlush into the
+// 						  incremental queue. The incrementalQueue worker will eventually
+// 						  add the results once its performed the incremental flush.
+func (s *peersSource) fetchBootstrapBlocksFromPeers(
+	shard uint32,
+	ranges xtime.Ranges,
+	nsMetadata namespace.Metadata,
+	session client.AdminSession,
+	bopts result.Options,
+	bootstrapResult result.BootstrapResult,
+	lock *sync.Mutex,
+	incremental bool,
+	incrementalQueue chan incrementalFlush,
+	shardRetrieverMgr block.DatabaseShardBlockRetrieverManager,
+	blockSize time.Duration,
+) {
+	it := ranges.Iter()
+	for it.Next() {
+		currRange := it.Value()
+
+		for start := currRange.Start; start.Before(currRange.End); start = start.Add(blockSize) {
+			version := s.opts.FetchBlocksMetadataEndpointVersion()
+			shardResult, err := session.FetchBootstrapBlocksFromPeers(nsMetadata,
+				shard, start, start.Add(blockSize), bopts, version)
```

Comment:
Perhaps add a `blockStart, blockEnd := start, start.Add(blockSize)` here so you can reuse here and when you specify `timeRange: xtime.Range{Start: blockStart, End: blockEnd}` you can reuse the var rather than doing the maths again (which isn't much overhead, but code more readable using the `blockStart`, `blockEnd` naming semantics.

Reply:
sure